File: company-name-matching/kotlin-examples/generate-match-report.kt
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

// Replace this with your API key from https://www.interzoid.com/manage-api-account
const val API_KEY = "YOUR_API_KEY_HERE"

// Input file containing one company name per line
const val INPUT_FILE_NAME = "sample-input-file.txt"

// Simple data class to hold the input company name and its similarity key (SimKey)
data class Record(
    val input: String,
    val simKey: String
)

fun main() {

    val records = mutableListOf<Record>()

    // Read the input file line by line
    val inputFile = File(INPUT_FILE_NAME)
    if (!inputFile.exists()) {
        println("Input file not found: $INPUT_FILE_NAME")
        return
    }

    inputFile.forEachLine { line ->
        val company = line.trim()

        // Skip blank lines
        if (company.isEmpty()) return@forEachLine

        val simKey = getCompanySimKey(company)

        // If no SimKey returned, skip this record
        if (simKey.isNullOrEmpty()) return@forEachLine

        records.add(Record(input = company, simKey = simKey))
    }

    if (records.isEmpty()) {
        println("No records with similarity keys found.")
        return
    }

    //----------------------------------------------------------------------
    // Sort records strictly by SimKey only so that matching SimKeys
    // are adjacent. This makes it easy to find clusters of matches.
    //----------------------------------------------------------------------
    records.sortBy { it.simKey }

    //----------------------------------------------------------------------
    // Walk through the sorted list and build clusters of records with
    // the same SimKey. Only print clusters containing two or more records.
    // Each printed line is formatted as "Input,SimKey" (two-column CSV)
    // and each cluster is separated by a blank line.
    //----------------------------------------------------------------------
    var currentKey: String? = null
    val cluster = mutableListOf<Record>()

    fun printCluster(c: List<Record>) {
        if (c.size < 2) return        // Only print clusters of size >= 2
        for (r in c) {
            println("${r.input},${r.simKey}")
        }
        println()                     // Blank line between clusters
    }

    for (rec in records) {
        if (rec.simKey != currentKey) {
            // New SimKey encountered: flush previous cluster
            if (cluster.isNotEmpty()) {
                printCluster(cluster)
                cluster.clear()
            }
            currentKey = rec.simKey
        }
        cluster.add(rec)
    }

    // Flush the last cluster at the end
    if (cluster.isNotEmpty()) {
        printCluster(cluster)
    }
}

/**
 * Calls Interzoid's getcompanymatchadvanced API for a single company name
 * and returns the similarity key (SimKey) as a String.
 * Returns null or empty if the call fails or no SimKey is present.
 */
fun getCompanySimKey(companyName: String): String? {
    // URL-encode the company so it is safe for use in a query parameter
    val companyParam = URLEncoder.encode(companyName, "UTF-8")

    // Build the API URL
    val apiUrl = "https://api.interzoid.com/getcompanymatchadvanced" +
            "?license=$API_KEY" +
            "&company=$companyParam" +
            "&algorithm=model-v4-wide"

    return try {
        val url = URL(apiUrl)
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"

        // Read the response body as a String
        val body = connection.inputStream.bufferedReader().use { it.readText() }

        connection.disconnect()

        // For this simple example, we avoid external JSON libraries and
        // manually extract the "SimKey" value from the JSON response.
        // The response is expected to look like:
        // {"SimKey":"...","Code":"Success","Credits":"..."}
        extractSimKey(body)
    } catch (e: Exception) {
        println("Error calling API for \"$companyName\": ${e.message}")
        null
    }
}

/**
 * Minimal helper to extract the value of the "SimKey" field
 * from a simple JSON object string like:
 * {"SimKey":"...","Code":"Success","Credits":"..."}
 *
 * This avoids adding external JSON dependencies for a small example.
 * For production code, use a proper JSON library such as Jackson or kotlinx.serialization.
 */
fun extractSimKey(json: String): String? {
    val key = "\"SimKey\""
    val keyIndex = json.indexOf(key)
    if (keyIndex == -1) return null

    // Find the first quote after "SimKey":
    val colonIndex = json.indexOf(':', keyIndex)
    if (colonIndex == -1) return null

    val firstQuote = json.indexOf('"', colonIndex + 1)
    if (firstQuote == -1) return null

    val secondQuote = json.indexOf('"', firstQuote + 1)
    if (secondQuote == -1) return null

    return json.substring(firstQuote + 1, secondQuote)
}

File: company-name-matching/node-examples/generate-match-report.js
// generate-match-report.js

const fs = require("fs");
const https = require("https");

// Replace this with your API key from https://www.interzoid.com/manage-api-account
const API_KEY = "YOUR_API_KEY_HERE";

// Input file containing one company name per line
const INPUT_FILE_NAME = "sample-input-file.txt";

/**
 * Calls Interzoid's getcompanymatchadvanced API for a single company name
 * and returns a Promise resolving to the similarity key (SimKey) as a string.
 * Returns an empty string on error or if SimKey is missing.
 */
function callCompanyMatchAPI(companyName) {
  return new Promise((resolve) => {
    // URL-encode the company name to safely embed it in the query string
    const companyParam = encodeURIComponent(companyName);

    const apiURL =
      "https://api.interzoid.com/getcompanymatchadvanced" +
      `?license=${API_KEY}` +
      `&company=${companyParam}` +
      "&algorithm=model-v4-wide";

    https
      .get(apiURL, (res) => {
        let data = "";

        res.on("data", (chunk) => {
          data += chunk;
        });

        res.on("end", () => {
          try {
            const json = JSON.parse(data);
            const simKey = json.SimKey || "";
            resolve(simKey);
          } catch (err) {
            console.error(
              `Error parsing JSON for "${companyName}": ${err.message}`
            );
            console.error("Raw response:", data);
            resolve("");
          }
        });
      })
      .on("error", (err) => {
        console.error(`Error calling API for "${companyName}": ${err.message}`);
        resolve("");
      });
  });
}

async function main() {
  // Each record will hold the original input and its similarity key
  const records = [];

  // Read the input file contents
  let fileContents;
  try {
    fileContents = fs.readFileSync(INPUT_FILE_NAME, "utf8");
  } catch (err) {
    console.error("Error reading input file:", err.message);
    return;
  }

  // Split into lines and process each non-empty line
  const lines = fileContents.split(/\r?\n/);

  for (const line of lines) {
    const company = line.trim();

    // Skip blank lines
    if (!company) continue;

    const simKey = await callCompanyMatchAPI(company);

    // Skip if no SimKey returned
    if (!simKey) continue;

    records.push({ input: company, simKey });
  }

  if (records.length === 0) {
    console.log("No records with similarity keys found.");
    return;
  }

  //--------------------------------------------------------------------
  // Sort records strictly by simKey only so that all matching keys
  // are adjacent in the array. This makes it easy to find clusters.
  //--------------------------------------------------------------------
  records.sort((a, b) => a.simKey.localeCompare(b.simKey));

  //--------------------------------------------------------------------
  // Walk the sorted list and build clusters of records that share
  // the same simKey. Only print clusters of size >= 2.
  // Each printed line is "Input,SimKey" (two-column CSV style),
  // and clusters are separated by a blank line.
  //--------------------------------------------------------------------
  let currentKey = null;
  let cluster = [];

  function printCluster(c) {
    if (c.length < 2) return; // Only print clusters with 2 or more
    for (const r of c) {
      console.log(`${r.input},${r.simKey}`);
    }
    console.log(); // blank line between clusters
  }

  for (const rec of records) {
    if (rec.simKey !== currentKey) {
      // New simKey: flush previous cluster
      if (cluster.length > 0) {
        printCluster(cluster);
      }
      currentKey = rec.simKey;
      cluster = [rec];
    } else {
      // Same simKey: add to current cluster
      cluster.push(rec);
    }
  }

  // Flush the final cluster
  if (cluster.length > 0) {
    printCluster(cluster);
  }
}

// Run the main function
main().catch((err) => {
  console.error("Unexpected error:", err);
});
